asap/asaplib/cluster/ml_cluster_fit.py

```diff
@@ -101,83 +104,280 @@ def pack(self):
+
+        # may need to re-think these criterion based on the type of density chosen
+
+        self.dens_cut = 0.2 * np.mean(np.log(self.dens)) + 0.8 * np.min(np.log(self.dens))
+        self.delta_cut = np.mean(self.distances)
+
+        self.delta = np.zeros(data.shape[0])  # shape = (Nele,)
+        self.ref = np.zeros(data.shape[0], dtype='int')  # shape = (Nele,)
+        tt = np.arange(data.shape[0])  # array from 0-Nele
+        imax = []
+        for i in range(data.shape[0]):
+            ll = tt[((self.dens > self.dens[i]) & (tt != i))]  # indices of data points with higher local density than x_i
+            dd = data[((self.dens > self.dens[i]) & (tt != i))]  # data points with higher local density than x_i
+            if dd.shape[0] > 0:  # If there is a data point with higher local density than x_i
+                ds = np.transpose(sp.spatial.distance.cdist([np.transpose(data[i, :])], dd))
+                j = np.argmin(ds)
+                self.ref[i] = ll[j]  # cluster centre for data point x_i
+                self.delta[i] = ds[j]  # distance to cluster centre for data point x_i
             else:
-                delta[i] = np.min(dist[i, js])
-                nneigh[i] = js[np.argmin(dist[i, js])]
-        return delta, nneigh
+                self.delta[i] =- 100.
+                imax.append(i)
+        self.delta[imax] = np.max(self.delta)*1.05
 
-    def pack(self):
-        '''return all the info'''
-        state = dict(deltamin=self.deltamin, rhomin=self.rhomin)
-        return state
+    def get_assignation(self, data):
+        """
+
+        Parameters
+        ----------
+        data: data matrix
+
+        Returns: center label: numpy array of shape (Nele,) giving the cluster label for a each data point.
+        -------
+
+        """
+        ordered = np.argsort(-self.dens)  # in descending order of density
+        self.cluster = np.zeros(data.shape[0], dtype='int')
+        tt = np.arange(data.shape[0])
+        center_label = np.zeros(data.shape[0], dtype='int')
+        ncluster = -1
+        for i in range(data.shape[0]):
+            j = ordered[i]
+            if (self.dens[j] > self.dens_cut) & (self.delta[j] > self.delta_cut):
+                ncluster = ncluster + 1
+                self.cluster[j] = ncluster
+                center_label[j] = ncluster
+            else:
+                self.cluster[j] = self.cluster[self.ref[j]]
+                center_label[j] = -1
+        self.centers = tt[(center_label != -1)]
+        bord = np.zeros(data.shape[0], dtype='int')
+        self.halo = np.copy(self.cluster)
+
+        for i in range(data.shape[0]):
+            for j in self.indices[i, :][(self.distances[i, :] <= self.dc)]:
+                if self.cluster[i] != self.cluster[j]:
+                    bord[i] = 1
+        halo_cutoff = np.zeros(ncluster + 1)
+        for i in range(ncluster + 1):
+            dd = self.dens[((bord == 1) & (self.cluster == i))]
```

Comment:
The current examples fail at this point where **dd** is returned as an empty array.

Reply:
Apologies, it's **examples/Ni-icosahedral-np** that I'm testing this with.